### ISSUE TITLE
possibility to run witcher 2 without GL_ARB_buffer_storage support

### ISSUE THREAD

**Issue by karolherbst** [virtual-programming/witcher2-linux#153]
currently I try to debug an issue witcher 2 has while running with nouveau on my system (https://bugs.freedesktop.org/show_bug.cgi?id=91310).

But to be able to debug this further, I need to be able to run witcher 2 without GL_ARB_buffer_storage support. Is there any way to disable the usage of this entirely?

Thanks a lot

**Comment by vpltd-peterm** [comment]
It should be enough to simply report that this extension is not available, and we fall back to an alternative code path using glBufferData() instead.

**Comment by karolherbst** [comment]
sadly I get crashes when I disable this extension through MESA_EXTENSION_OVERRIDE

**Comment by vpltd-peterm** [comment]
Here is a build of the GOG release of the game with ARB_buffer_storage explicitly disabled:
http://developer.vpltd.com/public/witcher2-gog-no_buffer_storage.tgz

This version also does not have the Google Breakpad crash handler enabled, which should aid debugging.

**Comment by karolherbst** [comment]
I have witcher only on steam and it fails with this in eon.txt

```[thread 00000001][W][97]: eON_LoadLibraryEx() couldn't load 'steam_api.dll', returning NULL!
[thread 00000001][E][97]: winpe_LinkImports() - Error loading DLL steam_api.dll
[thread 00000001][W][97]: DLL LOADING: Binding failed for C:\mnt\data\SteamLibrary\steamapps\common\the witcher 2\bin\\witcher2.exe at 0x400000, unloading module
[thread 00000001][W][97]: VirtualFree() fail: region is image region
[thread 00000001][W][97]: DLL LOADING: Binding failed for C:\mnt\data\SteamLibrary\steamapps\common\the witcher 2\bin\\witcher2.exe at 0x400000, module unloaded
[thread 00000001][E][97]: Loading Win32 executable failed!!
```

**Comment by licaon-kter** [comment]
Replace the executable with that one and run it.

**Comment by karolherbst** [comment]
@licaon-kter yeah, did exactly that, that's where the error is coming from

**Comment by vpltd-peterm** [comment]
I have uploaded the VPFS from the GOG release here:
http://developer.vpltd.com/public/witcher2-gog-vpfs.tgz

Replace the Steam witcher2.vpfs with this one, and there should be no problem. Of course, the Steam features such as achievements and Steam Cloud will stop working, but this should not be a problem for debugging :)

**Comment by vpltd-peterm** [comment]
Also Witcher 2 on Steam should give you a CD Key (usually it opens a popup about this). You may use this CD key at gog.com to redeem a full version from there

**Comment by karolherbst** [comment]
ohh okay, will try that

**Comment by karolherbst** [comment]
thanks, it works now. Strangly enough, with your binaries I can't reproduce the bug, but I don't have my old saves, how are the paths different between steam and gog, so that I can copy the saves around?

**Comment by vpltd-peterm** [comment]
Steam will have defaulted to using Steam Cloud by default. The saves are stored here:
$(HOME)/.steam/steam/userdata/<steam id>/20920/remote/

You should be able to move them to here:
$(HOME)/.local/share/cdprojektred/witcher2/GameDocuments/Witcher 2/gamesaves/

**Comment by karolherbst** [comment]
okay thanks, that worked. Now I have to figure out how to reproduce the bug with the new binary

**Comment by karolherbst** [comment]
@vpltd-peterm

just an idea: maybe the bug I encounter is already fixed somehow in your newer code base? I can't reproduce that with your binary and I somehow doubt that GL_ARB_buffer_storage is the cause of this. Would it be possible to get a new binary with GL_ARB_buffer_storage enabled like the one I got here already?

I just want to verify these idea and if a new binary with GL_ARB_buffer_storage enabled causes this issue, we can be pretty sure where the issue may be around there. (allthough intel and the blob didn't showed that issue, and I was sure it is a driver bug :/ )

**Comment by vpltd-peterm** [comment]
It's also quite possible that the bug was the result of some previously incorrectly generated shader on your system, and the root cause has been fixed. However, eON caches its generated shaders to $(HOME)/.local/share/cdprojektred/witcher2/eONPrecompiledShaders.dat, and its possible that this file contained a broken shader causing the problem.

The precompiled shader file has a "revision" field in it, which we bump from time to time. If the precompiled shader file is deemed too old, its contents are ignored and it is recreated from scratch. Deleting the precompiled shader file will also force eON to regenerate shaders. You may want to check if this was the cause of the problem initially.

If not, I will be happy to provide you with a new build of our current eON engine for this game, and you can advise us if the problem has been solved.

**Comment by karolherbst** [comment]
@vpltd-peterm

okay, thanks. I think this might be really an engine problem itself, because:

1. you new build without GL_ARB_buffer_storage doesn't show that bug to me and I doubt that GL_ARB_buffer_storage is the cause
2. when I load the arena first (with the steam build), the issue doesn't appear as well
3. gpu core speed effects the degree of the issue (higher gpu_core_freq/fps ratio => less artifacts), which means, when the engine has more time to draw the frame, less artifacts occur.

I would say there is some kind of syncing issue or anything like that, because the game behaves really strange in the main menu, best seen with vsync disabled. It seems to lock on 15 or 30 fps for some time, and then jumps up to 1000fps on 800x600 lowest settings without doing anything (noticed with the steam build, didn't check the gog build you gave me).

I think it would be best to just check the newest engine version and if that fixes the issue we can just forget about that (there was another bug about witcher, which got fixed, but without the one I encountered, and it may be the one used the current gog version? Sadly I need 6 hours to download witcher 2 from gog, so this would take some time to test.)

Thanks

**Comment by karolherbst** [comment]
sadly I didn't get the gog version to install, maybe due to lacking of disc space or /tmp to small or whatever, so I couldn't rally test this.

**Comment by karolherbst** [comment]
I think the current Steam version is still behind the GOG ones and it would be nice if there would be an update for this, so that there aren't those graphical issues anymore :)

**Comment by licaon-kter** [comment]
How do you know? Version numbers?

**Comment by karolherbst** [comment]
well it still shows the issues and not with the GOG binaries, but I will try out again to say it for sure.

**Comment by karolherbst** [comment]
@licaon-kter okay, I can confirm that with the GOG binaries linked above I don't have any issues (besides bad performance) on Nouveau. Is there a way how I can provide you with the version numbers from both binaries?

**Comment by licaon-kter** [comment]
In Steam it's in the Advance options, on gog.com I'm not sure, maybe the package has the version in it's name, or in some changelogs.

**Comment by karolherbst** [comment]
the Readme from the launcher says "The Witcher 2: Assassins of Kings Enhanced Edition ReadMe – June 18, 2012"

**Comment by licaon-kter** [comment]
Nope, that's not it.

**Comment by karolherbst** [comment]
So in Steam I only found the BuildID

765819 (linux_public_beta) and 607393 (normal) have the same issues, but that's everything I found. In the advanced options of the launcher I also didn't find anything.

**Comment by karolherbst** [comment]
@licaon-kter well in the .comment ELF section I found something, maybe it helps, maybe not: "GCC: (crosstool-NG 1.17.0) 4.6.3GCC: (Ubuntu 4.9.1-16ubuntu6) 4.9.1GCC: (Ubuntu 4.8.2-19ubuntu1) 4.8.2"

**Comment by licaon-kter** [comment]
Steam numbers are ok.
ELF section has no usable info.
Anyway someone like @vpltd-peterm or @vpltd-plesurf or vpltd-kgalaj might help.

**Comment by karolherbst** [comment]
thanks for your help though.

It would be just nice if the behaviour of both version would be the same and somebody told me, that it might be that the issues I encounter in the steam version are fixed in later engine updates and if an updated game binary fixes those graphical issues, then I don't see why mesa or Nouveau should hack around these ;)

**Comment by vpltd-kgalaj** [comment]
Hi. There's no way to run our games without GL_ARB_buffer_storage support. The necessity for this functionality stems directly from the differences between Direct3D and OpenGL, where ability of threads to issue OpenGL commands is concerned. In Direct3D9's device or Direct3D11's immediate context the only requirement is that no two threads can be inside Direct3D call to such object at the same time. In OpenGL on Linux (not on Mac), a thread is designated as the owner of OpenGL context, and from this point on until the moment when the thread itself relinquishes ownership no other thread can issue OpenGL commands to this context - at best they'll just get ignored, at worst there'll be a kernel panic if it tries. So, to adhere to DirectX's requirements (and more importantly, actual usage of DirectX by ported games) when simulating it with OpenGL on Linux, we have to keep a worker thread whose sole task is to issue OpenGL commands to OpenGL context, as they come through our internal command queue, and all the game's threads calling DirectX commands essentially place them in the queue, after as much preprocessing as possible. In case of OpenGL buffer mapping and buffer updates, after much evaluation through trial and error and time spend with profilers, it was determined that by far the quickest way to get these buffers available to OpenGL without additional copying is to use permanent mapping and buffer streaming. There is a fallback code to use when permanent mapping isn't available, but it isn't well maintained, mostly because we expect that without the ability to use permanent mapping to avoid having to pass buffer maps and updates to our OpenGL background thread the fps would go down by a factor of 10.

**Comment by karolherbst** [comment]
@vpltd-kgalaj I was asking about a version without GL_ARB_buffer_storage to be able to create an apitrace and debug the graphical issue, but the GOG version didn't had the issue, so we are at a totally different point now :/ Sorry for the confusion.

**Comment by karolherbst** [comment]
@vpltd-kgalaj by the way, this might also explain the bad performance I have with nouveau in saints row the third, because I get a lower performance by factor 6...

**Comment by vpltd-kgalaj** [comment]
@karolherbst I'm sorry; Peter is releasing new versions, so he's the best person to get hold of on what has changed between them. I'm mostly developing the engine and as bugs are fixed in it and new versions released (when we aren't barred from releasing any without extensive publisher's testing by original game's publisher's agreement), they make their way to the public. I wish I could focus on fixing all the bugs/performance issues in older releases (and new bugs that show up when other elements of the os are updated by their respective dev teams), but mostly I have to work on the new tech instead to keep the company afloat.

**Comment by karolherbst** [comment]
@vpltd-kgalaj yeah no problems. I was also wondering if anybody on your side ever tried to run games on nouveau (most likely not, because the reclocking code isn't stable enough yet, but if everything goes well, there will be stable kepler reclocking in 4.7 ... )

It just saddens me a bit, that we aren't really able to find those bottlenecks ourself yet (because we have to reverse engineer all that kind of stuff first), so we kind of rely on engine devs telling us where the critical paths are.

**Comment by vpltd-peterm** [comment]
Earlier in this thread (now a post by "ghost") I released a version of witcher 2 with ARB_buffer_storage disabled for the GOG version of the game. I did however make it clear that this was for debugging purposes only, and it should not be considered an "upgrade".

As far as I know, the official GOG and Steam releases should be in sync. If you find this to be different, please generate some sha1sum's of the executables for me and I'll see if i can find out what the changes are.

As @vpltd-kgalaj says, we can't guarantee correct functionality with this fallback path, though. It is largely going unmaintained now as ARB_buffer_storage is a much faster path.

Regarding testing on MESA drivers, it is true we do not do this regularly... the main reason for this is that we were discouraged early on by the need to opt into bleeding edge stuff which we frequently found unstable, leading to lost working hours trying to get systems running again.

**Comment by imirkin** [comment]
FWIW ARB_buffer_storage *is* supported by nouveau, and all the other [modern] mesa drivers. The issue is that if you use coherent mappings, that defeats apitrace, which is often used for debugging. Persistent but non-coherent mappings still work (i.e. things that require explicit flush).

Also from a general performance standpoint, non-coherent persistent mappings are more performant since the driver doesn't have to flush caches on every draw if there has been no change.

**Comment by vpltd-peterm** [comment]
Krystian would be better to answer this, but I seem to recall that during our testing this was not the case - at least not with the nVidia and AMD binary drivers

**Comment by imirkin** [comment]
Depending on the specifics, the performance difference could be minimal. (e.g. if you update all your persistently-mapped buffers between every draw, you're not saving on any cache flushes... or if your datasets blow through the caches anyways, losing that data could not be a big deal, etc.)

**Comment by karolherbst** [comment]
@vpltd-peterm I will try to get the GOG version installed here and try it out with this one.

**Comment by vpltd-peterm** [comment]
If you have the Steam version, the only difference is the executable and the VPFS file. If you have these from the links I gave above, then that's all you need :)

If you do download the GOG version, don't try to simply extract the data from it. There is a file which is split into 3 parts, and needs to be joined together by the postinstall script.

**Comment by karolherbst** [comment]
@vpltd-peterm okay, with the files from the links above I don't get any graphical issues, but with the steam ones I do.

my sha1sums:
```
a53c3adda13a11bfb76606ff628912930c017a12  GOG/witcher2
6d15eb1feb975516142e4c91b1da16e85b3513ed  GOG/witcher2.vpfs
2292832fc4b29ea232d5be3224664a40b0717541  STEAM/witcher2
afd50eeacd69bbb027613dcaea3d48682229c74d  STEAM/witcher2.vpfs
```

**Comment by licaon-kter** [comment]
That would be the Steam 607393 (normal).
The Steam 765819 (linux_public_beta) release has:
```
df0e275d0b35dd1b65220703521b817982db6a0e  STEAM/witcher2
afd50eeacd69bbb027613dcaea3d48682229c74d  STEAM/witcher2.vpfs
```

**Comment by karolherbst** [comment]
on the linux_public_beta release I get a different one for the vpfs file:
```
df0e275d0b35dd1b65220703521b817982db6a0e  witcher2
6d15eb1feb975516142e4c91b1da16e85b3513ed  witcher2.vpfs
```

**Comment by karolherbst** [comment]
ohh wait, I still left the vpfs from gog there, nevermind then

**Comment by licaon-kter** [comment]
You can edit/delete comments here :)

**Comment by karolherbst** [comment]
right

**Comment by karolherbst** [comment]
Maybe I should just test the current GOG binary, but I really don't want to download the 20GB if I haven't to.

Anyway, to sum it up: both steam version get me graphical issues. The gog binaries from the link above, does not.

![screenshot_20160416_210858](https://cloud.githubusercontent.com/assets/171766/14583091/971dbe7a-0417-11e6-8f5e-b3b4683461fa.png)

**Comment by vpltd-peterm** [comment]
There could be another issue here - when you run the game from Steam, are you running it under the Steam runtime ?

You can test this with the GOG version by launching it using ~/.steam/bin32/steam-runtime/run.sh

**Comment by karolherbst** [comment]
@vpltd-peterm nope. I use my system provided libraries for both. But I also start it through steam, because I simply overwrite the files inside the directory

**Comment by karolherbst** [comment]
@vpltd-peterm okay. I just got the same issue with saints row 2 and disabling GL_ARB_buffer_storage fixes those issues. So I think it is this in the end and nouveau just does something stupid. I will still disable it and try with witcher2 and report back

**Comment by vpltd-kgalaj** [comment]
@imirkin I did make both coherent and non-coherent path when I was first expanding the engine to cover persistent mapping, switchable by a #define in DX9 module code. The non-coherent path however performed on NVidia drivers much worse than coherent one, and when I reached to NVidia for guidance I was essentially told to just use the coherent path, since non-coherent one reduced to coherent one in driver code but in more complex way. A lot of things may have changed since then, but I'd need to be able to focus on it for longer to re-do proper testing now, as I'm not sure if the #define is enough for the game to just run after defining it.... but just to test... Peter, if you want and can, you could make a non-coherent build with DX9 ( the USE_MEMORY_BARRIER_DIRECTLY define ), but the DX11 code only has coherent path atm.

**Comment by karolherbst** [comment]
@vpltd-kgalaj 
okay, with a hack I was able to run witcher 2 and saints row 2  with GL_ARB_buffer_storage  enabled on nouveau without any graphical issues. But I lack the proper knowledge to write a proper fix and because it seems I am the only nouveau developer with a valid license for saints row 2 I would like to ask if it is possible to get a developer key/license/whatever so that another nouveau dev can properly debug this.

On a side note: the issue is already showing in the saints row 2 main menu screen. Maybe it would be also possible to have a small demo?

Thanks a lot.

**Comment by licaon-kter** [comment]
http://store.steampowered.com/promotion/familysharing ?

**Comment by karolherbst** [comment]
@licaon-kter for that both need physical access to the same machine

**Comment by imirkin** [comment]
The main point is whether VP is interested in having nouveau developers take a look at their games, and attempt to make them work properly. If they are - great, then keys shouldn't be a problem. If not, then that's fine too, and we'll keep the VP-made ports on the "known to not work" list. (There is, of course, no guarantee that they would be made to work, but it's vastly more likely that someone will take a look if they have access to the game.)

**Comment by vpltd-kgalaj** [comment]
I think that nearly all games VP ported aren't really VP's, that is, the company paying for the port keeps all access to keys. So to dispense any keys to nouveau devs VP would first have to buy them in a normal way, or ask the relevant company for them. I'll ask Paul.

**Comment by licaon-kter** [comment]
@karolherbst Umm, no, you just need to be their friend on the list an mark that user as your family, afaics, to be able to share (lend) him the game.

**Comment by karolherbst** [comment]
@licaon-kter nope. There are two things which are important for sharing a library:

1. the computer has to be activated for the library for being used (up to 10 computers allowed)
2. You can make your library accesible through another account (upt to 5 accounts).

but 1. always applies in any situation. I do family sharing already, and I encountered this.

**Comment by karolherbst** [comment]
@vpltd-kgalaj ohh. I would have expected that there are some sort of keys, which could be given to driver developers, because it is usually in the intereset of the game publisher having a well performing driver for that game :/

Oh well, a bit sad that the situation is as you said.

**Comment by vpltd-kgalaj** [comment]
@karolherbst Ok, we have two codes, one for SR2 and one for SR3. Obviously I can't post them here... would the contact email I see on your github profile be the right place to send them to?

**Comment by karolherbst** [comment]
@vpltd-kgalaj yeah, that would be awesome, then we can discuss internally who should get it or how we make it accesible to all devs or something. Thanks a lot :)

**Comment by vpltd-kgalaj** [comment]
Well, attempt to send an email to git@karolherbst.de results in information "domain not found"...

**Comment by karolherbst** [comment]
mhh there seems something wrong with the domain, I will check that.

in the meantime you can use karolherbst@gmail.com

sorry for the troubles

**Comment by vpltd-kgalaj** [comment]
Sent :)

**Comment by vpltd-peterm** [comment]
For future reference, legit requests from MESA developers for access to our games - contact me at peter.mulholland@vpltd.com or make contact with "jaycee1980" on freenode #dri-devel
